### PR TITLE
chore: use shillelagh instead of gsheetsdb

### PR DESCRIPTION
**Pull request by betodealmeida**
### SUMMARY

[shillelagh](https://github.com/betodealmeida/shillelagh/) is a Python module that provides a SQLAlchemy dialect and DB API 2.0 implementation to a variety of non-SQL resources. For example, it can be used to query CSV files or fetch weather data from https://www.weatherapi.com/ via SQL. The library has 100% unit test coverage and uses SQLite virtual tables to allow efficiently accessing the resources as if they were tables in a database.

The current module that we use to access Google Spreadsheets is [gsheetsdb](https://github.com/betodealmeida/gsheets-db-api/), which depends on the unmaintained [moz-sql-parser](https://github.com/mozilla/moz-sql-parser). In addition to the dependency, `gsheetsdb` implements a code transpiler, and uses pre and post-processors to fetch data from the spreadsheet, making the code somewhat brittle.

This week I implemented [a Google Spreadsheets adapter in shillelagh](https://github.com/betodealmeida/shillelagh/blob/master/src/shillelagh/adapters/api/gsheets.py), aiming to deprecate the `gsheetsdb` module. The implementation is a drop-in replacement, and users should be able to uninstall the `gsheetsdb` module and install the `shillelagh` module to have existing queries, datasets and charts continue working.

This works because `shillelagh` [registers a SQLAlchemy dialect called "gsheets"](https://github.com/betodealmeida/shillelagh/blob/master/setup.cfg#L83). This way SQLAlchemy will use `shillelagh` for all engines with the URL "gsheets://". This is an [in-memory only backend](https://github.com/betodealmeida/shillelagh/blob/e53bd116d5ffb2f8a99f29ad3913ab50300d373e/src/shillelagh/backends/apsw/dialect.py#L146), without access to the filesystem, so it's safe to use.

On the other hand, `shillelagh` also registers its own dialect (aptly named "shillelagh", with an alias called "shillelagh+apsw"). That dialect is unsafe, as a malicious user could use it to read data from the filesystem by enabling the CSV adapter. For security reasons, I disabled the "shillelagh" dialect unless `PREVENT_UNSAFE_DB_CONNECTIONS` is on.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

N/A

### TEST PLAN

Tested SQL Editor and charts reading from Google spreadsheets, everything works.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by betodealmeida** [comment]
While fixing unit tests I did some clean up and removed duplicated code and duplicated exceptions.

File: setup.py [hunk 1]
```diff
@@ -129,7 +129,7 @@ def get_git_sha():
         "elasticsearch": ["elasticsearch-dbapi>=0.2.0, <0.3.0"],
         "exasol": ["sqlalchemy-exasol>=2.1.0, <2.2"],
         "excel": ["xlrd>=1.2.0, <1.3"],
-        "gsheets": ["gsheetsdb>=0.1.9"],
+        "gsheets": ["shillelagh>=0.2"],
```
Comment (villebro) [inline setup.py:132]: I would add `, <0.3` here, just in case.

File: superset/views/core.py [hunk 2]
```diff
@@ -1234,7 +1231,7 @@ def testconn(  # pylint: disable=too-many-return-statements,no-self-use
         uri = request.json.get("uri")
         try:
             if app.config["PREVENT_UNSAFE_DB_CONNECTIONS"]:
-                check_sqlalchemy_uri(uri)
+                check_sqlalchemy_uri(make_url(uri))
```
Comment (villebro) [inline superset/views/core.py:1234]: I'm surprised we weren't getting mypy errors here before

**Review by villebro — APPROVED** [review]
One minor nit - this is a great step forward for as the `gsheets` connector seems to be fairly widely used. Also big thanks for cleaning up the related code 👍

File: superset/views/core.py (hunk 2 above)
Comment (betodealmeida) [reply superset/views/core.py:1234]: Me too.

File: setup.py (hunk 1 above)
Comment (betodealmeida) [reply setup.py:132]: +1

**Pull request merged**